### PR TITLE
connections-pane: handle postgres materialized views

### PR DESCRIPTION
**Pull request by detule**
Closes https://github.com/r-dbi/odbc/issues/255

Thought about going couple of different ways here - including masquerading the materialized views as simple "view" s.  In the end opted for leaving them marked as they are and handling "matview" in the self reported "objtype" structure.

Pros/cons to both - happy to discuss and/or go in a different direction.

### CODE REVIEW

**Review by simonpcouch — COMMENTED** [review]
I don't have much intuition here, so bear with me if this question isn't well-defined.😬

From a quick internet search, it seems like materialized views are supported by quite a few DBMS'. Do we want to design this in such a way where it will "just work" for new DBMS' without us having to add new methods (and thus not have to introduce the `connection` argument in the first place)? i.e. starting with `validateObjectName()` as-is on main, add `...` to the signature and pass them in from `odbcListColumns()`, and then inline what's currently proposed as the PostgreSQL method as: 

```r
matview <- list(...)$matview
view <- view %||% matview
```

File: R/connection-pane.R [hunk 1]
```diff
@@ -38,11 +38,18 @@ odbcListObjectTypes.default <- function(connection) {
...
-  }
+  # See if we have views too.  A little more elaborate than just
+  # checking if any( table_types == "VIEW" ), because some back-ends
+  # may contain VIEW "look-alikes".  For example PostgreSQL/"MATVIEW".
+  # The icon url is that of the default Rstudio IDE resource location.
+  # It gets automatically pulled in for an object type "view" ( but
+  # not for some of the other ones ).  Should we not want to encode that
+  # dependency, we can just bring that PNG into this package.
+  table_types <- tolower(string_values(odbcConnectionTableTypes(connection)))
+  viewlike <- table_types[grepl("view", table_types)]
```
[the remark this answers is not in the document]
Comment (detule) [reply R/connection-pane.R:49]: Yep / i like it.  Thanks!

File: R/connection-pane.R [hunk 2]
```diff
@@ -38,11 +38,18 @@ odbcListObjectTypes.default <- function(connection) {
...
+  viewlike_types <-
+    sapply(viewlike, function(i) list(contains = "data", icon="connections/objects/view.png"), simplify=FALSE)
```
[the remark this answers is not in the document]
Comment (detule) [reply R/connection-pane.R]: Done / thanks!

**Comment by detule** [comment]
Hey Simon - thanks for taking a look.

Good suggestion and I am happy to go in that direction.  I made it a generic because, maybe incorrectly, I thought that if supported in back-ends other than PSQL materialized views would be listed as something other than "matview".  I checked in Oracle, and there they are returned as "TABLE".  Can you check databricks, and/or Snowflake ( apparently my non-enterprise account is too pedestrian to support materialized views ) ?

Thanks again!

**Comment by simonpcouch** [comment]
> too pedestrian to support materialized views

😆

> Can you check databricks, and/or Snowflake...

Ah, I see! Thanks. Yeah, in Snowflake the materialized view is listed as a view. Able to `odbcListColumns()` just fine but returns an error when attempting to `odbcPreviewObject()` (from the connections pane).

**Comment by detule** [comment]
> > Can you check databricks, and/or Snowflake...
> 
> Ah, I see! Thanks. Yeah, in Snowflake the materialized view is listed as a view. Able to `odbcListColumns()` just fine but returns an error when attempting to `odbcPreviewObject()` (from the connections pane).

OK - looks like listed as "matview" in PSQL, "table" in Oracle, "view" in Snowflake.  Not much of a theme there - but I think also no need to customize similar to what we did with for PSQL.   Going with your suggestion of doing away with the generic.  Let me know what you think.

We can fix the snowflake preview fail as a follow up.

File: R/connection-pane.R [hunk 3]
```diff
@@ -184,8 +194,15 @@ computeDisplayName <- function(connection) {
...
 }
 
-# selects the table or view from arguments
-validateObjectName <- function(table, view) {
+validateObjectName <- function(table, view, ...) {
+
+  # Handle view look-alike object types
+  # ( e.g. PostgreSQL/"matview" )
+  args <- list(...)
+  argNames <- names(args)
```
Comment (simonpcouch) [inline R/connection-pane.R:202]:
```suggestion
  arg_names <- names(args)
```

File: R/connection-pane.R [hunk 4]
```diff
@@ -184,8 +194,15 @@ computeDisplayName <- function(connection) {
...
+  viewlike <- grep("view", argNames, value = TRUE)
```
Comment (simonpcouch) [inline R/connection-pane.R:203]:
```suggestion
  viewlike <- grep("view", arg_names, value = TRUE)
```

**Review by simonpcouch — APPROVED** [review]
Apologies for leaving this PR in an ambiguous place—thumbs-up from me. :)

**Pull request merged**